linkerd/meshtls/verifier/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+        .map_err(|error| tracing::warn!(%error, "Failed to extract tls id from client end cert"))
+        .ok()
+        .unwrap_or_default();
+
+    if ids.contains(expected_id) {
+        return Ok(());
+    }
+
+    Err(io::Error::new(
+        io::ErrorKind::Other,
+        "certificate does not match TLS identity",
+    ))
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::verify_id;
+    use linkerd_identity as id;
+    use rcgen::generate_simple_self_signed;
+    use std::str::FromStr;
+
+    fn generate_cert_with_names(names: Vec<&str>) -> Vec<u8> {
+        let sans: Vec<String> = names.into_iter().map(|s| s.into()).collect();
+
+        generate_simple_self_signed(sans)
+            .expect("should generate cert")
+            .serialize_der()
+            .expect("should serialize")
+    }
+
+    #[test]
+    pub fn cert_with_dns_san_matches_dns_id() {
+        let dns_name = "foo.ns1.serviceaccount.identity.linkerd.cluster.local";
+        let cert = generate_cert_with_names(vec![dns_name]);
+        let id = id::Id::from_str(dns_name).expect("should parse DNS id");
+        assert!(verify_id(&cert, &id).is_ok());
+    }
+
+    #[test]
+    pub fn cert_with_dns_san_does_not_match_dns_id() {
+        let dns_name_cert = vec!["foo.ns1.serviceaccount.identity.linkerd.cluster.local"];
+        let dns_name = "bar.ns1.serviceaccount.identity.linkerd.cluster.local";
+
+        let cert = generate_cert_with_names(dns_name_cert);
+        let id = id::Id::from_str(dns_name).expect("should parse DNS id");
+        assert!(verify_id(&cert, &id).is_err());
+    }
+
+    #[test]
+    pub fn cert_with_uri_san_does_not_match_dns_id() {
+        let uri_name_cert = vec!["spiffe://some-trust-comain/some-system/some-component"];
+        let dns_name = "bar.ns1.serviceaccount.identity.linkerd.cluster.local";
+
+        let cert = generate_cert_with_names(uri_name_cert);
+        let id = id::Id::from_str(dns_name).expect("should parse DNS id");
+        assert!(verify_id(&cert, &id).is_err());
+    }
+
+    #[test]
+    pub fn cert_with_no_san_does_not_verify_for_dns_id() {
+        let dns_name = "bar.ns1.serviceaccount.identity.linkerd.cluster.local";
+        let cert = generate_cert_with_names(vec![]);
+        let id = id::Id::from_str(dns_name).expect("should parse DNS id");
+        assert!(verify_id(&cert, &id).is_err());
+    }
+
+    #[test]
+    pub fn cert_with_dns_multiple_sans_one_matches_dns_id() {
+        let foo_dns_id = "foo.ns1.serviceaccount.identity.linkerd.cluster.local";
+        let bar_dns_id = "bar.ns1.serviceaccount.identity.linkerd.cluster.local";
+        let spiffe_id = "spiffe://some-trust-comain/some-system/some-component";
+
+        let cert = generate_cert_with_names(vec![foo_dns_id, bar_dns_id, spiffe_id]);
+        let id = id::Id::from_str(foo_dns_id).expect("should parse DNS id");
+        assert!(verify_id(&cert, &id).is_ok());
+    }
+
+    #[test]
+    pub fn cert_with_dns_multiple_sans_none_matches_dns_id() {
```

Comment:
Is it worth adding a test with a cert that has only a spiffe_id? I.e. to document that we return an error (until the next PR).

Reply:
If I understand you correctly, we do have such a test: https://github.com/linkerd/linkerd2-proxy/pull/2534/files#diff-f01d3fc9f7b0d20d3a9c55e350ffb752b18e784f08731b61a0afda65f9c09477R88